File: tasks/createMissingTrainers.js
const fs = require('fs');
const path = require('path');
const parentFilePath = path.resolve(__dirname, '..');
const TRAINER_TABLE = require(path.join(parentFilePath, 'input', 'TrainerTable.json'));
const {getTrainerLabel, getTrainerName, getTrainerDataFromPlaceDatas} = require('./trainerUtils');

const BDSPTrainers = getTrainerDataFromPlaceDatas();

function findMissingTrainers(trainers) {
    const arr = [...Array(2000).keys()];
    arr.splice(0, 1);
    for(let trainer of trainers) {
        const idx = arr.findIndex(e => e === trainer.trainerId);
        arr.splice(idx, 1);
    }

    let volkners = [];
    let missingTrainers = [];
    for(let element of arr) {
        const trainerData = TRAINER_TABLE.TrainerData[element];
        const trainerType = TRAINER_TABLE.TrainerType[trainerData.TypeID];
        const trainerLabel = getTrainerLabel(trainerType.LabelTrType);
        const trainerName = getTrainerName(trainerData.NameLabel);

        if(trainerName === 'Volkner') {
            //We extended the trainer by duplicating Volkner trainers, therefore any Volkners outside of the docs' specified IDs should be considered undefined trainers
            volkners.push({ trainerId: element, name: trainerName, type: trainerLabel})
        } else {
            missingTrainers.push({ trainerId: element, name: trainerName, type: trainerLabel});
        }
    }

   return {
    volkners,
    missingTrainers
   }
}

module.exports = function() {
    const {missingTrainers} = findMissingTrainers(BDSPTrainers);
    fs.writeFileSync(path.join(parentFilePath, 'output', 'missingTrainers.json'), JSON.stringify(missingTrainers), 'utf-8');
};

File: python/createMissingTrainers.py
import os
import json
from trainerUtils import get_trainer_label, get_trainer_name, get_trainer_data_from_place_datas

parent_file_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
trainertable_file_path = os.path.join(parent_file_path, 'input', 'TrainerTable.json')

BDSP_Trainers = get_trainer_data_from_place_datas()
TRAINER_TABLE = {}

with open(trainertable_file_path, 'r') as f:
    TRAINER_TABLE = json.load(f)

def find_missing_trainers(trainers):
    arr = list(range(1, 2000))
    for trainer in trainers:
        try:
            arr.remove(trainer['trainerId'])
        except ValueError:
            pass

    volkners = []
    missing_trainers = []
    for element in arr:
        trainer_data = TRAINER_TABLE['TrainerData'][element]
        trainer_type = TRAINER_TABLE['TrainerType'][trainer_data['TypeID']]
        trainer_label = get_trainer_label(trainer_type['LabelTrType'])
        trainer_name = get_trainer_name(trainer_data['NameLabel'])

        if trainer_name == 'Volkner':
            # We extended the trainer by duplicating Volkner trainers, therefore any Volkners outside of the docs' specified IDs should be considered undefined trainers
            volkners.append({'trainerId': element, 'name': trainer_name, 'type': trainer_label})
        else:
            missing_trainers.append({'trainerId': element, 'name': trainer_name, 'type': trainer_label})

    return {'volkners': volkners, 'missingTrainers': missing_trainers}

result = find_missing_trainers(BDSP_Trainers)
with open(os.path.join(parent_file_path, 'output', 'missingTrainers.json'), 'w') as f:
    json.dump(result['missingTrainers'], f, indent=2)